teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -0,0 +1,113 @@
+- Decouple api and feature versioning by updating existing beta features now.
+- Decouple api and feature versioning by updating existing beta features after a 9 month window.
+- Make changes to only new beta features.
+
+All migration plans aim to decouple api and feature versioning, but they differ in the time window given to users to adapt to the backwards incompatible changes and the scope of the changes (i.e. whether they are made to existing features or new features only).
+
+### Decouple api and feature versioning now
+Require `enable-api-fields` to be set to `beta` when using beta features, regardless of CRD API version. Immediately make this change for existing beta features. This will solve the unintended behavior that taskruns cannot be created for version coupled features with “enable-api-fields” set to “stable” during the [v1 storage swap](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707) right away.
+
+## Design Details
+
+### Update validation for existing beta features now
+Currently, there are validations for beta feature in v1 apis, for example [validation for resolver in resolverRef validation](https://github.com/tektoncd/pipeline/blob/0b95e49b707798ce2b19d29f31bd1ae82bebe3ee/pkg/apis/pipeline/v1/taskref_validation.go#L39) and [valiation for object params in task validation](https://github.com/tektoncd/pipeline/blob/0b95e49b707798ce2b19d29f31bd1ae82bebe3ee/pkg/apis/pipeline/v1/task_validation.go#L112). However, these validations are missing in v1beta1 apis, which has contributed to the fact that beta features have been regarded as stable in v1beta1, thus leading to the coupling of feature versioning and apiVersion. 
+
+This TEP proposes to update validations for exisiting beta feature immediately given that this has led to a bug where users are not able to create a v1 TaskRun if they are creating a Task with beta featurs with CRD specified as v1beta1. And after the change, beta features in v1beta1 would also require `enable-api-fields` to be set to `beta` to use.
+
+### Update api compatibility policy
+This TEP also proposes to update the [api compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) to reflect that in order to turn on beta features, it would require the `enable-api-fields` to be set to `beta` regardless of the CRD apiVersion. New features that are promoted to beta stability will also follow this policy, regardless of the CRD's API version. It will clarify that feature stability levels are not correlated with CRD API versions.
+
+### Deprecate the default `beta` stability level
+This TEP proposes that once we are able to decouple api and feature versioning, we should switch back to `stable` as the default `enable-api-fields` value. This would allow us to deprecate the current default value for opt-in features of `beta` stability level.
```

Comment:
What's the reason to move back to enable-api-fields = stable by default? This would be another breaking change.

We previously had enable-api-fields = stable and changed it to beta, but this didn't change the fact that beta features were enabled by default for the v1beta1 api. This would make beta features disabled by default for the v1beta1 API. I don't see a strong reason to make this change, and I also don't think this decision needs to happen as part of this TEP.

Reply:
I agree with @JeromeJu that we need to switch back the flag default to "stable" - having the default set to "beta" was a workaround for an issue generated by the decoupling, but it's not something we should keep.
Unfortunately, as you noted, it is a breaking change, but only for API features that are currently in beta. 
What we should do is deprecate the default value being beta, as @JeromeJu suggested, and after 9 months, or as soon as those beta features are stable (whichever happens first) switch the default to "stable".

I think the deprecation itself should be part of this TEP, but the switch to stable does not need to be, we can create an issue for use to remember about it.

Reply:
@afrittoli can you elaborate on your thinking re: why "stable" is a more appropriate default than "beta"? It sounds like you prefer this approach because the default was previously "stable", but I'd note that for the v1beta1 API, beta features have always been on by default, and this change would turn beta features off by default.

I'm also a bit confused about what it means to deprecate the default value of a flag, rather than to deprecate the flag itself. Since we're not removing the flag and cluster operators can change its value at any time, what's the benefit of announcing a change 9 months in advance? If the deprecation is part of this TEP, doesn't that also imply that this TEP includes a decision to change the default value to stable?

I think the questions of "what's the most appropriate value for enable-api-fields, and is it worth making a breaking change to set enable-api-fields to what we believe is the most appropriate value?" can be tackled separately from this TEP.

Reply:
Thanks for the discussion on this @afrittoli  and @lbernick both offline and through the comments! 🙏 

 I think first we all agree that with what this TEP proposes, we would be able to default `stable` as the `enable-api-fields` since it was because of the decoupling of api and feature versioning we had to switch to `beta` as the default `enable-api-fields`. But how and when to do so remains to be discussed, and I have also learned the following points:
- We need to decide what is the value for `enable-api-fields` in the long run for pipelines.
- Whether we are going to make this as a backward incompatible change or not remains a question since we need to  gather info from users and cluster operators standpoints on whether giving notice really benefits them.

Therefore, I have opened https://github.com/tektoncd/pipeline/issues/6948. And I am hoping to reach a conclusion there on what's the next step after we decouple the api and feature versioning once TEP0138 is implemented.